[css-anchor-position-1] Tighten the state management in AnchorPositionedState

anttijk opened the pull request:
#### d246b8e72c4fb421054c6ac396c7e79e27537ca9
<pre>
[css-anchor-position-1] Tighten the state management in AnchorPositionedState
<a href="https://bugs.webkit.org/show_bug.cgi?id=289163">https://bugs.webkit.org/show_bug.cgi?id=289163</a>
<a href="https://rdar.apple.com/problem/146313542">rdar://problem/146313542</a>

Reviewed by Sam Weinig.

Currently we do too many style updates for anchor positioned elements because transient state is not properly managed.
This is inefficient and also hides bugs (which this patch fixes).

The patch makes the stage enum an optional paramater that is only initialized if we are actually resolving anchors.
This is a step toward moving the transient state to a TreeResolver scoped struct.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-transition-eval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-try-order-position-area-expected.txt:

These were spurious PASSes as try-order evaluation is not implemented.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveStyle):

Scope Style::Resolver lifetime so it gets deleted before compositing layer update.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::setSnapshottedScrollOffsetForAnchorPositioning):
(WebCore::RenderLayer::clearSnapshottedScrollOffsetForAnchorPositioning):

Ensure we do compositing layer update if offset change.

* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::findAnchorForAnchorFunctionAndAttemptResolution):
(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositionedStateForLayoutTimePositioned):

Update if it might be layout time positioned (has position-anchor);

* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::invalidateForAnchorDependencies):

Only save the anchor rect if anchored elements have been found.

(WebCore::Style::Scope::resetAnchorPositioningStateBeforeStyleResolution):

Clear the stage.

(WebCore::Style::Scope::updateAnchorPositioningStateAfterStyleResolution):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolve):

Don&apos;t trigger further style resolution passes if the stage has not progressed.

(WebCore::Style::TreeResolver::updateAnchorPositioningState):

If an anchor changes style we need to do an interleaving round since its position may change.

(WebCore::Style::TreeResolver::generatePositionOptionsIfNeeded):
(WebCore::Style::TreeResolver::tryChoosePositionOption):

Don&apos;t pick an option before it has been positioned.

Canonical link: <a href="https://commits.webkit.org/291656@main">https://commits.webkit.org/291656@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5b5b37bd1c072981425ec4176a7879adbd760219

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93564 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13134 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2868 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98567 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44090 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95614 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13423 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21580 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71479 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28855 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96566 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10035 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84609 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51814 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9717 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2232 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43404 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79992 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2291 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100599 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20616 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15076 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80492 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20868 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80544 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79826 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19856 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24361 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1708 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13764 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20600 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25778 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20287 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23747 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22028 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/21aca17dd15ee69fc13da150bfcc832d8e39a129. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41929

webkit-commit-queue commented:
Commit message contains (OOPS!), blocking PR #41929. Details: [Build #19840](https://ews-build.webkit.org/#/builders/19/builds/19840)

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5b5b37bd1c072981425ec4176a7879adbd760219. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41929

anttijk committed:
[css-anchor-position-1] Tighten the state management in AnchorPositionedState
https://bugs.webkit.org/show_bug.cgi?id=289163
rdar://problem/146313542

Reviewed by Sam Weinig.

Currently we do too many style updates for anchor positioned elements because transient state is not properly managed.
This is inefficient and also hides bugs (which this patch fixes).

The patch makes the stage enum an optional paramater that is only initialized if we are actually resolving anchors.
This is a step toward moving the transient state to a TreeResolver scoped struct.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-transition-eval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-try-order-position-area-expected.txt:

These were spurious PASSes as try-order evaluation is not implemented.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveStyle):

Scope Style::Resolver lifetime so it gets deleted before compositing layer update.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::setSnapshottedScrollOffsetForAnchorPositioning):
(WebCore::RenderLayer::clearSnapshottedScrollOffsetForAnchorPositioning):

Ensure we do compositing layer update if offset change.

* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::findAnchorForAnchorFunctionAndAttemptResolution):
(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositionedStateForLayoutTimePositioned):

Update if it might be layout time positioned (has position-anchor);

* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::invalidateForAnchorDependencies):

Only save the anchor rect if anchored elements have been found.

(WebCore::Style::Scope::resetAnchorPositioningStateBeforeStyleResolution):

Clear the stage.

(WebCore::Style::Scope::updateAnchorPositioningStateAfterStyleResolution):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolve):

Don't trigger further style resolution passes if the stage has not progressed.

(WebCore::Style::TreeResolver::updateAnchorPositioningState):

If an anchor changes style we need to do an interleaving round since its position may change.

(WebCore::Style::TreeResolver::generatePositionOptionsIfNeeded):
(WebCore::Style::TreeResolver::tryChoosePositionOption):

Don't pick an option before it has been positioned.

Canonical link: https://commits.webkit.org/291656@main

webkit-commit-queue commented:
Committed 291656@main (d246b8e72c4f): <https://commits.webkit.org/291656@main>

Reviewed commits have been landed. Closing PR #41929 and removing active labels.
